### ISSUE TITLE
`DenseAssetStorage#insert` "unreachable" panic

### ISSUE THREAD

**Issue by JeanMertz** [bevyengine/bevy#10444]
## Bevy version

`0.12.0`

## What you did

I don't currently have an MRE, but wanted to call out my small headless app with a custom asset loader, hitting this line after upgrading from 0.11 to 0.12. I assume I botched the migration and will go through that when I have the time, but even so, the fact that this is marked as `unreachable` tells me either Bevy has a leaky abstraction, or this should be changed to a `panic` (possibly with a more helpful error message).

https://github.com/bevyengine/bevy/blob/4852fc7578b7a55db4c05a6fb9f8ebdd290571e1/crates/bevy_asset/src/assets.rs#L149

**Comment by JeanMertz** [comment]
Quick follow-up:

I have a call to `Assets#remove` somewhere, which causes this panic in 0.12. Changing it to `Assets#get` and then cloning the result works as expected.

**Comment by JeanMertz** [comment]
So, I _think_ what was happening is that I had a system that would run `asset_server.load("...")` and then `my_assets.get(&handle)` _until_ that call would return `Some`. The examples show doing the `load` once and then storing the asset, I never bothered, since it worked without doing the store+load step.

Since upgrading to 0.12, that started causing issues. I've since switched to using a `Local<Option<Handle<MyAsset>>>`, and doing:

```rust
let handle = my_local.get_or_insert_with(|| asset_server.load("..."));
let Some(FooAsset(foo)) = my_assets.get(&*handle) else {
    return;
};
```

And now things are working as expected again.

**Comment by JeanMertz** [comment]
Actually, even with the above change, although `my_assets.get` now works as expected, `my_assets.remove` still results in the same panic as in the original issue description.

**Comment by nicopap** [comment]
I'm getting this panic even when using `asset_server.load("…")` only once for my custom asset. Also I'm fairly certain I never call `Assets::remove`. Investigating.

**Comment by LennysLounge** [comment]
I stumbled on this panic when first removing an asset with `Assets::remove` and the reloading the asset with `AssetServer::reload`. 
This causes the panic every time.

Funny enough, first calling `Assets::remove` and then `AssetServer::load` does not cause the panic.

**Comment by LennysLounge** [comment]
For what its worth here is an example how to cause this panic:
```rs
use bevy::prelude::*;

fn main() {
    App::new()
        .add_plugins(DefaultPlugins)
        .add_systems(Startup, setup)
        .add_systems(Update, asset_added)
        .run();
}
fn setup(mut commands: Commands, asset_server: Res<AssetServer>) {
    commands.spawn(SpriteBundle {
        texture: asset_server.load("bevy.png"),
        ..Default::default()
    });
}

fn asset_added(
    asset_server: Res<AssetServer>,
    mut assets: ResMut<Assets<Image>>,
    mut asset_events: EventReader<AssetEvent<Image>>,
) {
    for event in asset_events.read() {
        if let AssetEvent::<Image>::Added { id } = event {
            assets.remove(id.clone());
            asset_server.reload("bevy.png");
        }
    }
}
```

I did some digging why this happens.   
Like the panic says, entries should always be valid. An entry becomes invalid when it is removed from the asset storage and its index is queued up to be recycled. When a new index is reserver, an old index may be recycled the next time flush is called. However when reloading an asset, it doesnt reserve a new index, it keeps using the index that was removed. This causes the index to never be actually recycled and it remains invalid. Ergo the panic.

Two things to note:  
1) When an asset is removed from an `Asstes` it will trigger a `AssetEvent::Removed` event however the asset server still reports its load state as `Loaded`. I seems weird for an asset to have been removed and to be loaded at the same time. Perhaps this should return the load state to `NotLoaded` in the asset server?

2) As described in the docs, `AssetServer::reload` should only reload assets that are in the `Loaded` state. However since the load state isnt quite correct this causes the panic. Should reloading a "removed" asset be a thing at all?

**Comment by marcelchampagne** [comment]
I'm also getting this panic when upgrading from Bevy 0.11 to 0.12. I don't have any custom assets and I've yet to track down what in my code could be causing this. The only thing I call on the AssetServer is asset_server.load("..."), I've never used reload or remove.

**Comment by kristoff3r** [comment]
I tried debugging this a bit by running the example from @LennysLounge and inserting some extra logging into bevy:

```
2023-11-22T23:14:10.201238Z  INFO bevy_asset::assets: flush: 0 -> 1
2023-11-22T23:14:10.201267Z  INFO bevy_asset::assets: Inserting at AssetIndex { generation: 0, index: 0 }: [true]
2023-11-22T23:14:10.214301Z  INFO bevy_asset::assets: flush: 1 -> 2
2023-11-22T23:14:10.214310Z  INFO bevy_asset::assets: Inserting at AssetIndex { generation: 0, index: 1 }: [true, true]
2023-11-22T23:14:10.237964Z  INFO bevy_asset::assets: flush: 2 -> 3
2023-11-22T23:14:10.237973Z  INFO bevy_asset::assets: Inserting at AssetIndex { generation: 0, index: 2 }: [true, true, true]
2023-11-22T23:14:10.394400Z  INFO bevy_diagnostic::system_information_diagnostics_plugin::internal: SystemInfo { os: "Linux 23.05 NixOS", kernel: "6.1.62", cpu: "AMD Ryzen 7 5800X 8-Core Processor", core_count: "8", memory: "47.0 GiB" }
2023-11-22T23:14:10.413256Z  INFO bevy_asset::assets: flush: 3 -> 4
2023-11-22T23:14:10.413274Z  INFO bevy_asset::assets: Inserting at AssetIndex { generation: 0, index: 3 }: [true, true, true, true]
2023-11-22T23:14:10.413561Z  INFO bevy_asset::assets: flush: 4 -> 4
2023-11-22T23:14:10.413578Z  INFO bevy_asset::server: Reloading bevy.png because it has changed
2023-11-22T23:14:10.413592Z  INFO bevy_asset::assets: flush: 4 -> 4
2023-11-22T23:14:10.413588Z  INFO bevy_asset::server: Reloading bevy.png because it has changed
2023-11-22T23:14:10.413610Z  INFO bevy_asset::assets: flush: 4 -> 4
2023-11-22T23:14:10.413610Z  INFO bevy_asset::server: Reloading bevy.png because it has changed
2023-11-22T23:14:10.413639Z  INFO bevy_asset::server: Reloading bevy.png because it has changed
2023-11-22T23:14:10.492179Z  INFO bevy_asset::assets: flush: 4 -> 4
2023-11-22T23:14:10.492203Z  INFO bevy_asset::assets: Inserting at AssetIndex { generation: 0, index: 3 }: [false, false, false, true]
2023-11-22T23:14:10.492233Z  INFO bevy_asset::assets: flush: 4 -> 4
2023-11-22T23:14:10.492238Z  INFO bevy_asset::assets: Inserting at AssetIndex { generation: 0, index: 3 }: [false, false, false, true]
2023-11-22T23:14:10.492247Z  INFO bevy_asset::assets: flush: 4 -> 4
2023-11-22T23:14:10.492252Z  INFO bevy_asset::assets: Inserting at AssetIndex { generation: 0, index: 3 }: [false, false, false, true]
2023-11-22T23:14:10.492260Z  INFO bevy_asset::assets: flush: 4 -> 4
2023-11-22T23:14:10.492264Z  INFO bevy_asset::assets: Inserting at AssetIndex { generation: 0, index: 3 }: [false, false, false, true]
2023-11-22T23:14:10.492736Z  INFO bevy_asset::assets: flush: 4 -> 4
2023-11-22T23:14:10.492747Z  INFO bevy_asset::server: Reloading bevy.png because it has changed
2023-11-22T23:14:10.517362Z  INFO bevy_asset::assets: flush: 4 -> 4
2023-11-22T23:14:10.517383Z  INFO bevy_asset::assets: Inserting at AssetIndex { generation: 0, index: 3 }: [false, false, false, false]
```

And this is the `flush` function for context (with my logging added):
```
    pub(crate) fn flush(&mut self) {
        // NOTE: this assumes the allocator index is monotonically increasing.
        let new_len = self
            .allocator
            .next_index
            .load(std::sync::atomic::Ordering::Relaxed);

        info!("flush: {} -> {}", self.storage.len(), new_len);
        self.storage.resize_with(new_len as usize, || Entry::Some {
            value: None,
            generation: 0,
        });
        while let Ok(recycled) = self.allocator.recycled_receiver.try_recv() {
            info!("Recycling index {}", recycled.index);
            let entry = &mut self.storage[recycled.index as usize];
            *entry = Entry::Some {
                value: None,
                generation: recycled.generation,
            };
        }
    }
```

There are multiple things I don't understand here:
* It seems to be allocating the same asset 4 times and then stopping, presumably because it reloads before it finishes loading? That seems like a (different) bug.
*  The invariant that gets hit after `flush` assumes that we can find a free storage index, because either we have allocated something new (so we resize the storage), or we have reclaimed something from `recycled_receiver`. But this is trivially broken by the `remove` method on `Assets`, as it removes storage while doing neither of these. It seems like it needs to be moved from `Assets` over to `AssetServer`, then it can also handle setting the `AssetLoaded` as @LennysLounge noted above.

As an aside I think having some more logging in the new asset system would help a lot, both for debugging stuff like this but also for communicating the intent of the code and for understanding what happens when using the system.

**Comment by cart** [comment]
@kristoff3r 
> It seems to be allocating the same asset 4 times and then stopping, presumably because it reloads before it finishes loading? That seems like a (different) bug.

This is happening because you are listening for all events and there are 3 other image assets created (by bevy render iirc) and then kicking off a reload of the `bevy.png` asset for each one.

It fails on the 4th because that is the one where you actually remove the `bevy.png` asset

**Comment by cart** [comment]
I think this bug is pretty straightforward and largely comes down to a missed case (re-inserting at an id that has been removed and re-loaded). This is an interesting case because the handle is "still alive" and the asset server also considers it to be alive (which is why the reload is allowed to go through).

I'm thinking theres going to be a reasonable way to handle this case. I just need to think for a bit about how we _want_ this to behave (Ex: do we allow this insert to go through, restoring the removed asset, or do we disallow this case and consider invalidating the handle / updating its state).

**Comment by cart** [comment]
A big thanks to everyone here for investigating (and to @LennysLounge for providing a minimal repro example). You made it very easy to hop in!

**Comment by cart** [comment]
There are cases to consider here, such as ensuring this scenario will work: https://github.com/bevyengine/bevy/pull/10520

**Comment by JMS55** [comment]
@cart in #10520 what I did was add a new `AssetEvent::NoLongerUsed` that's emitted when the last strong handle is dropped. I also found an existing bug where the track_assets() system assumes the asset wasn't removed yet when the last handle is dropped.

In general, the AssetServer needs to be revamped a bit to deal with assets than have been removed already.

**Comment by marcelchampagne** [comment]
@cart I'm still encountering the same panic when attempting to upgrade to 0.12.1. I spent a couple hours getting a MRE and I've created a new issue here: https://github.com/bevyengine/bevy/issues/10820.

Thanks in advance!
